staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go

```diff
@@ -900,11 +900,19 @@ func (c *Cacher) processEvent(event *watchCacheEvent) {
 	c.incoming <- *event
 }
 
-func (c *Cacher) dispatchEvents() {
+func (c *Cacher) dispatchEvents(progressRequester *conditionalProgressRequester) {
 	// Jitter to help level out any aggregate load.
 	bookmarkTimer := c.clock.NewTimer(wait.Jitter(time.Second, 0.25))
 	defer bookmarkTimer.Stop()
 
+	// TODO(p0lyn0mial): adapt the following logic once
+	//  https://github.com/kubernetes/kubernetes/pull/124612 merges
+	progressRequesterCleanUpOnceFn := func() { /*no-op*/ }
+	if utilfeature.DefaultFeatureGate.Enabled(features.ConsistentListFromCache) && utilfeature.DefaultFeatureGate.Enabled(features.WatchList) {
```

Comment:
maybe we should actually gate the progressRequester only when the etcd version "matches" ?

Reply:
Yes - we should gate on `DefaultFeatureSupportsChecker` once that merges.
The problem there is that it may not yet be initialized... and we need to handle that case too, so it's a bit more tricky (because we are not able to differentiate between not-initialized and not-supports really).
@serathius - FYI

[On a related note, we should probably reject streaming list requests if progress-requester is not supported.

Reply:
> because we are not able to differentiate between not-initialized and not-supports really

why ? It is not supported when the version of etcd doesn't match.

> On a related note, we should probably reject streaming list requests if progress-requester is not supported.

yeah, ideally if we could add `etcdVersionChecker` to https://github.com/kubernetes/kubernetes/blob/2ae115e5a2c9b0ce3be8b3f4b60f8acc1b7c9108/staging/src/k8s.io/apimachinery/pkg/apis/meta/internalversion/validation/validation.go#L28

could the `etcdVersionChecker` gate the server readiness until it initialises ?

Reply:
> why ? It is not supported when the version of etcd doesn't match.

etcd can be started after kube-apiserver, so we default to false and let initialization switch it

> could the etcdVersionChecker gate the server readiness until it initialises ?

no - because we don't really know when it is fully initialized...

Reply:
> etcd can be started after kube-apiserver, so we default to false and let initialization switch it

In that case the server won't be ready until [newETCD3Check](https://github.com/kubernetes/kubernetes/blob/3e5b03eb433ee359782f5aa6e9368ab2a0d0370c/staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go#L152) turns green. We could create something similar for the version checker.